### ISSUE TITLE
ASSERTION FAILED: The bit sizes of src and the dst is not identical on Haswell GT2 Mobile

### ISSUE THREAD

**Issue by linnaea** [linnaea/beignet#1]
Tracking for upstream GitLab 81.

Issue seems to be that structs with 3 elements that looks like vec3 being converted to vec3, while CL spec says vec3 should take 4 elements' space rather than 3.

```c
typedef struct __attribute__ ((packed)) {
    uint cols;
    uint diagl;
    uint diagr;
} start;

#define DEPTH 9
#define WG_SIZE 64

typedef short int_fast16_t;
typedef ushort uint_fast16_t;
typedef uint uint_fast32_t;
#define UINT_FAST32_MAX UINT_MAX

// for convenience
#define L (get_local_id(0))
#define G (get_global_id(0))

kernel void solve(__global const start* in_starts, __global uint* out_cnt) {
    uint num = 0;
    uint_fast32_t cols[WG_SIZE][DEPTH+1], posibs[WG_SIZE][DEPTH+1];
    uint_fast32_t diagl[WG_SIZE][DEPTH+1], diagr[WG_SIZE][DEPTH+1];
    int_fast16_t d = 1;
    cols[L][d] = in_starts[G].cols;
    diagl[L][d] = in_starts[G].diagl;
    diagr[L][d] = in_starts[G].diagr;
    uint_fast32_t posib = ( diagl[L][d] | diagr[L][d]);
    uint_fast32_t bit = ~posib & (posib + 1);
    uint_fast32_t new_diagr = (bit >> 1) | diagr[L][d];
    bit |= cols[L][d];
    uint_fast32_t new_posib = (bit | new_diagr);

    if (new_posib != UINT_FAST32_MAX) {
      cols[L][d] = bit;
    } else {
        num += bit == UINT_FAST32_MAX;
    }
    posib = posibs[L][d];
    out_cnt[G] += num;
}
```

**Comment by linnaea** [comment]
Repro
```c
typedef struct __attribute__ ((packed)) {
    uint a;
    uint b;
    uint c;
} s;

kernel void k(__global const s* ins, __global uint* out) {
  uint a = ins[0].a;
  uint b = ins[0].b;
  uint c = ins[0].c;
  out[0] = a;
  out[1] = b;
  out[2] = c;
}
```

Gives the following sequence
```
  %0 = bitcast %struct.s addrspace(1)* %ins to <3 x i32> addrspace(1)*
  %1 = load <3 x i32>, <3 x i32> addrspace(1)* %0, align 1
  %2 = extractelement <3 x i32> %1, i32 0
  %3 = extractelement <3 x i32> %1, i32 1
  %4 = extractelement <3 x i32> %1, i32 2
  %5 = bitcast i32 addrspace(1)* %out to <3 x i32> addrspace(1)*
  %6 = insertelement <3 x i32> undef, i32 %2, i32 0
  %7 = insertelement <3 x i32> %6, i32 %3, i32 1
  %8 = insertelement <3 x i32> %7, i32 %4, i32 2
  store <3 x i32> %8, <3 x i32> addrspace(1)* %5, align 4
  ret void
```

**Comment by linnaea** [comment]
Should be fixed in e576632
